proposals/254-padding-negotiation.txt

```diff
@@ -436,42 +170,100 @@ increasing bin width allows the histograms to most accurately represent
-state's histogram, then whenever a padding packet is sent, the
-corresponding histogram bin's token count is decremented by one.
+Tokens can be optionally removed from histogram bins whenever a padding or
+non-padding packet is sent. With this token removal, the histogram functions
+as an overall target delay distribution for the machine while it is in that
+state.
 
-If a packet matching the current state's transition_reschedule_events
-bitmask arrives from the server before the chosen padding timer expires,
-then a token is removed from a non-empty bin corresponding to
-the delay since the last packet was sent, and the padding packet timer
-is re-sampled from the histogram.
+If token removal is enabled, when a padding packet is sent, a token is removed
+from the bin corresponding to the target delay. When a non-padding packet is
+sent, the actual delay from the previous packet is calculated, and the
+histogram bin corresponding to that delay is inspected.  If that bin has
+tokens remaining, it is decremented.
+
+If the bin has no tokens left, the state removes a token from a different bin,
+as specified in its token removal rule. The following token removal options
+are defined:
+  * None -- Never remove any tokens
+  * Exact -- Only remove from the target bin, if it is empty, ignore it.
+  * Higher -- Remove from the next higher non-empty bin
+  * Lower -- Remove from the next higher non-empty bin
+  * Closest -- Remove from the closest non-empty bin by index
+  * Closest_time -- Remove from the closest non-empty bin by index, by time
+
+When all bins are empty in a histogram, the padding machine emits the internal
+"bins empty" event to itself.
+
+3.2. Machine Selection
+
+Clients will select which of the defined available padding machines to use
+based on the conditions that these machines specify. These conditions include:
+  * How many hops the circuit must be in order for the machine to apply
+  * If the machine requires vanguards to be enabled to apply
+  * The state the circuit must be in for machines to apply (building,
+    relay early cells remaining, opened, streams currently attached).
+  * If the circuit purpose matches a set of purposes for the machine.
+  * If the target hop of the machine supports circuit padding.
+
+Clients will only select machines whose conditions fully match given circuits.
+
+3.3. Machine Neogitation
+
+When a machine is selected, the client uses leaky-pipe delivery to send a
+RELAY_COMMAND_PADDING_NEGOTIATE to the target hop of the machine, using the
+following trunnel relay cell payload format:
+
+  /**
+   * This command tells the relay to alter its min and max netflow
+   * timeout range values, and send padding at that rate (resuming
+   * if stopped). */
+  struct circpad_negotiate {
+    u8 version IN [0];
+    u8 command IN [CIRCPAD_COMMAND_START, CIRCPAD_COMMAND_STOP];
+  
+    /** Machine type is left unbounded because we can specify
+     * new machines in the consensus */
+    u8 machine_type;
```

Comment:
Maybe we can make this u8 a bit bigger? Might save us a version update at some point.